demos/fleet_provisioning/fleet_provisioning_with_csr/pkcs11_operations.c

```diff
@@ -0,0 +1,1162 @@
+    /* Unreferenced parameters. */
+    ( void ) ( pContext );
+    ( void ) ( pRng );
+    ( void ) ( pRngContext );
+    ( void ) ( mdAlg );
+
+    /* Sanity check buffer length. */
+    if( hashLen > sizeof( toBeSigned ) )
+    {
+        ret = CKR_ARGUMENTS_BAD;
+    }
+
+    mech.mechanism = CKM_ECDSA;
+    memcpy( toBeSigned, pHash, hashLen );
+    toBeSignedLen = hashLen;
+
+    if( ret == CKR_OK )
+    {
+        ret = C_GetFunctionList( &functionList );
+    }
+
+    if( ret == CKR_OK )
+    {
+        ret = functionList->C_SignInit( signingContext.p11Session, &mech,
+                                        signingContext.p11PrivateKey );
+    }
+
+    if( ret == CKR_OK )
+    {
+        *pSigLen = sizeof( toBeSigned );
+        ret = functionList->C_Sign( signingContext.p11Session, toBeSigned,
+                                    toBeSignedLen, pSig, ( CK_ULONG_PTR ) pSigLen );
+    }
+
+    if( ret == CKR_OK )
+    {
+        /* PKCS #11 for P256 returns a 64-byte signature with 32 bytes for R and 32
+         * bytes for S. This must be converted to an ASN.1 encoded array. */
+        if( *pSigLen != pkcs11ECDSA_P256_SIGNATURE_LENGTH )
+        {
+            ret = CKR_FUNCTION_FAILED;
+            LogError( ( "Failed to sign message using PKCS #11. Expected signature "
+                        "length of %lu, but received %lu.",
+                        ( unsigned long ) pkcs11ECDSA_P256_SIGNATURE_LENGTH,
+                        ( unsigned long ) *pSigLen ) );
+        }
+
+        if( ret == CKR_OK )
+        {
+            PKI_pkcs11SignatureTombedTLSSignature( pSig, pSigLen );
+        }
+    }
+
+    if( ret != CKR_OK )
+    {
+        LogError( ( "Failed to sign message using PKCS #11 with error code %lu.", ret ) );
+        result = -1;
+    }
+
+    return result;
+}
+
+/*-----------------------------------------------------------*/
+
+static int randomCallback( void * pCtx,
+                           unsigned char * pRandom,
+                           size_t randomLength )
+{
+    CK_SESSION_HANDLE * p11Session = ( CK_SESSION_HANDLE * ) pCtx;
+    CK_RV res;
+    CK_FUNCTION_LIST_PTR pP11FunctionList;
+
+    res = C_GetFunctionList( &pP11FunctionList );
+
+    if( res != CKR_OK )
+    {
+        LogError( ( "Failed to generate a random number in RNG callback. Could not get a "
+                    "PKCS #11 function pointer." ) );
+    }
+    else
+    {
+        res = pP11FunctionList->C_GenerateRandom( *p11Session, pRandom, randomLength );
+
+        if( res != CKR_OK )
+        {
+            LogError( ( "Failed to generate a random number in RNG callback. "
+                        "C_GenerateRandom failed with %lu.", res ) );
+        }
+    }
+
+    return ( int ) res;
+}
+
+/*-----------------------------------------------------------*/
+
+static CK_RV generateKeyPairEC( CK_SESSION_HANDLE session,
+                                uint8_t * privateKeyLabel,
+                                uint8_t * publicKeyLabel,
```

Comment:
Labels can be passed as `const char *` type parameters for all functions

Reply:
We can do so but then we have to cast them wherever used as PKCS #11 APIs take them non-const.